### DISCUSSION TITLE
Proposal 4: Web 7.0 TSLF Proposal as a Picture: Separate (but parallel) Inbound and Outbound Message Processing Stacks

### DISCUSSION THREAD

**Discussion by mwherman2000** [trustoverip/trust-spanning-protocol#21] [Ideas]
Partly inspired by Daniel's proposal slide 8 as well as the Web 7.0 InterCredential Verifiable Credential Sender-Receiver Model, I believe the more correct model needs to have Separate (but parallel) Inbound and Outbound Message Processing Stacks (as illustrated in the diagram below).  

The primary issue with slide 8 is that it assumes "bottom up" processing (with optional components/steps).  Bottoms-up is only appropriate for inbound message processing for messages to be received by a (DIDComm) Receiver Agent.  It also doesn't address attachments (as far as I can tell). 

This one-way message transfer model has two roles: a Sender and a Receiver. The actions of the Sender move down the TSL framework; the actions of a Receiver move up the the TSL framework. The Sender and Receiver action are obviously related but they are not step-for-step symmetric.

What about the (DIDComm) Sender Agent? I believe this question contorts slide 8 enough to warrant a counter proposal: Proposal 3: Web 7.0 TSLF Proposal (this proposal).

**UPDATED: Proposal 3: Web 7.0 TSLF Proposal as a Picture: Separate (but parallel) Inbound and Outbound Message Processing Stacks**
![image](https://user-images.githubusercontent.com/6101736/219925020-f3c474b0-73a3-402f-8016-fa81e50a37a4.png)

**Web 7.0 InterCredential Verifiable Credential Sender-Receiver Model**
![image](https://user-images.githubusercontent.com/6101736/219360014-d9e92346-ade8-402f-ab32-8107dcf863d9.png)
Reference: https://www.youtube.com/watch?v=pvqkrt9q1ew&list=PLU-rWqHm5p44AsU5GLsc1bHC7P8zolAAf&index=6&t=0s

**Daniel's Slide 8**
![image](https://user-images.githubusercontent.com/6101736/219360152-d712191a-ad43-452c-8a35-2c23fce88ddb.png)
Reference: https://github.com/trustoverip/trust-spanning-protocol/discussions/19

**Comment by dhh1128** [comment]
Just a clarification, @mwherman2000 . I have ALWAYS assumed that all communication must be framed as simplex or one-way insofar as we are imagining its capabilities. That doesn't mean that two-way comms don't happen; it means that how a message gets from Alice to Bob is a totally separate question from how a message gets from Bob to Alice. It is possible that a response flows back over the same route and the same channels, but it is also possible that it flows back a totally different way. This was a design requirement of DIDComm and I simply carried it over without thinking and without clarifying -- so it was undoubtedly not obvious from my slides or presentation. But that is the intent.

My picture is a stack in the abstract, for one given direction of flow only. If a message flows the other direction, then it would be a stack that has the same logical structure -- but because of my unspoken design requirement, I never imagined that we could make the assumption that they were actually the same stack instantiation.

That's a long way of saying that I strongly agree with you that inbound and outbound must be separable. It is possible that in a given circumstances, they could be conflated, but that conflation is not a generally safe assumption.

**Comment by mwherman2000** [comment]
@dhh1128 My post has nothing to do with bi-directional communication (aka two-way exchange of messages). 

A one-way message transfer model has two _roles_: a Sender and a Receiver.  The actions of the Sender move down the TSL framework; the actions of a Receiver move up the the TSL framework.  Apologies for being so abrupt.

> My picture is a stack in the abstract, for one given direction of flow only.

So is mine, that is: _for one given direction of flow only_.  Mine is not abstract. _Need to love software developers differently_.  ;-) :-) 

p.s. I added to a third Message Transport arrow to the Proposal 3 diagram to help prevent this confudsion in the future.

**Comment by mwherman2000** [comment]
@dhh1128 Daniel, if you scrutinize the 8 Sender actions and the 6 Receiver actions in Proposal 3: Web 7.0 TSLF Proposal, they provide pretty good if not 100% coverage of the 5 categories in your Proposal 1.

**UPDATED: Proposal 3: Web 7.0 TSLF Proposal as a Picture: Separate (but parallel) Inbound and Outbound Message Processing Stacks**
![image](https://user-images.githubusercontent.com/6101736/219925109-77ad2902-dad7-48d4-bd9c-350a2d280329.png)

**Comment by mwherman2000** [comment]
Good footnote to this conversation that appears to be dying on the vine: https://twitter.com/mwherman2000/status/1628042135378223105?t=iZMtKENRoRwhU8IYFJ6UHA&s=19

**Comment by mwherman2000** [comment]
@dhh1128 @SmithSamuelM side note: I know you and Sam are pushing hard to KERIze the ToIP Trust Spanning Layer model but I want to drop a word of caution here: don't be so over zealous that we end up with what the JSON-LD/RDF Block of partners did to the W3C VC2 data model spec: made only useable if you force yourself to use @context and @vocab to serve no useful purpose other than serve as useless Christmas tree decorations.
Allow for a truly layered model that developers will understand...like the one I've proposed above.

**Comment by talltree** [comment]
@mwherman2000 Can I ask a favor? You titled this discussion "Proposal 3" (and also this other discussion you started [here](https://github.com/trustoverip/trust-spanning-protocol/discussions/23)), which of course seems logical since we've had two formal proposals so far. However we announced at last week's meeting that Proposal #3 would be the one Wenjing Chu is giving in tomorrow's TSPTF meeting. So to avoid confusion, could you rename this discussion and the other one you started to "Proposal #4"?

**Reply by mwherman2000** [reply]
> So to avoid confusion, could you rename this discussion and the other one you started to "Proposal #4"?

...grumble...  Where's the Discussion presentation of Wenjing Chu's?  "Proposal https://github.com/trustoverip/trust-spanning-protocol/pull/3" isn't real.

**Reply by talltree** [reply]
Thanks Michael. Wenjing just posted his discussion thread [here](https://github.com/trustoverip/trust-spanning-protocol/discussions/25). He'll add more links after his presentation to the TSPTF tomorrow.

**Reply by mwherman2000** [reply]
> Thanks Michael. Wenjing just posted his discussion thread [here](https://github.com/trustoverip/trust-spanning-protocol/discussions/25). He'll add more links after his presentation to the TSPTF tomorrow.

I should have used Proposal 7 😉🙂

**Comment by mwherman2000** [comment]
The formal summary presentation can be found here: https://github.com/trustoverip/trust-spanning-protocol/discussions/27
